qemu/tests/qemu_guest_agent.py

```diff
@@ -877,6 +878,146 @@ def gagent_check_fsfreeze(self, test, params, env):
+            """
+            As file related cmd is in blacklist at default,so need to change.
+            Now only linux guest has this behavior,but still leave interface
+            for windows guest.
+            """
+            if params.get("os_type") == "linux":
+                output = session.cmd_output(params["cmd_verify"])
+                if output == "":
+                    logging.info("Guest-file related cmds are already in white list.")
+                    return
+
+                session.cmd(params["cmd_change_black"])
+                output = session.cmd_output(params["cmd_verify"])
+                if not output == "":
+                    test.fail("Failed to change guest-file related cmd to white list, "
+                              "the output is %s" % output)
+
+                status, output = session.cmd_status_output(params["gagent_restart_cmd"])
+                if status:
+                    test.fail("Could not restart qemu-ga in VM, detail: %s" % output)
+                else:
+                    logging.info("Succeed to change guest-file related cmd "
+                                 "to white list.")
+
+        session = self._get_session(params, self.vm)
+        self._open_session_list.append(session)
+
+        error_context.context("Change guest-file related cmd to white list.")
+        change_bl()
+
+        error_context.context("Read guest file with file-read cmd.",
+                              logging.info)
+        # read guest file which are already created in guest.
+        logging.info("Create file in guest.")
+        ranstr = utils_misc.generate_random_string(5)
+        tmp_file = "%s%s" % (params["file_path"], ranstr)
+        content = "hello world"
+        session.cmd("echo %s > %s" % (content, tmp_file))
+
+        logging.info("Open and read the guest file.")
+        ret_handle = int(self.gagent.file_open(tmp_file, "r"))
+        read(ret_handle, content)
+
+        logging.info("Close the opened file.")
+        self.gagent.file_close(ret_handle)
+
+        # create a new file by agent.
+        error_context.context("Create new file with mode 'w' and write to it",
+                              logging.info)
+        ranstr = utils_misc.generate_random_string(5)
+        tmp_file = "%s%s" % (params["file_path"], ranstr)
+        ret_handle = int(self.gagent.file_open(tmp_file, "w+"))
+
+        logging.info("Write content to the file.")
+        content = "%s\n" % content
+
+        self.gagent.file_write(ret_handle, content)
+
+        logging.info("check file content from guest.")
+        cmd_check_file = "%s %s" % (params["cmd_read"], tmp_file)
+        status, output = session.cmd_status_output(cmd_check_file)
+        if status:
+            test.error("Read file content cmd failed, the output is %s." % output)
+        elif output == "":
+            logging.info("Content is in cache, need flush it.")
+            self.gagent.file_flush(ret_handle)
```

Comment:
@xiagao  if output isn't "", then flush command won't be executed. If so, it will miss this command.  we should cover this command each test.

Reply:
@xiangchunfu   Like we talked in private,I have resolved three issues you comment.
Pls have a check again,  thanks.